### DISCUSSION TITLE
Can I use cuBLAS directly in AMReX? (or are there alternatives?)

### DISCUSSION THREAD

**Discussion by antcarreon** [AMReX-Codes/amrex#3603] [Q&A] [answered]
Howdy everyone,

I'm relatively new to AMReX. I want to implement a simple matrix-matrix multiplication function using cuBLAS within AMReX; however, I have concerns about memory management since this is usually done automatically with arenas. Otherwise, what alternatives are there for implementing GPU-accelerated matrix-matrix multiplication (using Nvidia GPUs in particular)? Thank you for the help!

**Comment by WeiqunZhang** [comment, marked as answer]
You should be able to use cuBLAS. As for memory management, you can replace
```
double* p;
cudaMalloc(&p, size_in_bytes);
dudaFree(p);
```
with
```
auto* p = (double*) The_Arena()->alloc(size_in_bytes);
The_Arena()->free(p);
```
